DICOM/src/Microsoft.Health.Dicom.Anonymizer.Core/AnonymizerEngine.cs

```diff
@@ -0,0 +1,83 @@
+// -------------------------------------------------------------------------------------------------
+
+using Dicom;
+using EnsureThat;
+using Microsoft.Extensions.Logging;
+using Microsoft.Health.Dicom.Anonymizer.Core.AnonymizerConfigurations;
+using Microsoft.Health.Dicom.Anonymizer.Core.Model;
+using Microsoft.Health.Dicom.Anonymizer.Core.Processors.Settings;
+using Microsoft.Health.Dicom.Anonymizer.Core.Rules;
+
+namespace Microsoft.Health.Dicom.Anonymizer.Core
+{
+    public class AnonymizerEngine
+    {
+        private readonly ILogger _logger = AnonymizerLogging.CreateLogger<AnonymizerEngine>();
+        private readonly AnonymizerSettings _anonymizerSettings;
+        private readonly AnonymizerConfigurationManager _configurationManager;
+        private readonly IAnonymizerRuleFactory _ruleFactory;
+
+        public AnonymizerEngine(string configFilePath = "configuration.json", AnonymizerSettings anonymizerSettings = null, IAnonymizerRuleFactory ruleFactory = null)
+            : this(AnonymizerConfigurationManager.CreateFromConfigurationFile(configFilePath), anonymizerSettings, ruleFactory)
+        {
+        }
+
+        public AnonymizerEngine(AnonymizerConfigurationManager configurationManager, AnonymizerSettings anonymizerSettings = null, IAnonymizerRuleFactory ruleFactory = null)
+        {
+            EnsureArg.IsNotNull(configurationManager, nameof(configurationManager));
+
+            _anonymizerSettings = anonymizerSettings ?? new AnonymizerSettings();
+            _configurationManager = configurationManager;
+            _ruleFactory = ruleFactory ?? new AnonymizerRuleFactory(_configurationManager.Configuration, new DicomProcessorFactory(), new AnonymizerSettingsFactory());
+            _logger.LogDebug("Successfully initialized anonymizer engine.");
+        }
+
+        public void AnonymizeDataset(DicomDataset dataset)
+        {
+            EnsureArg.IsNotNull(dataset, nameof(dataset));
+
+            var context = InitContext(dataset);
+            ValidateInput(dataset);
+            dataset.AutoValidate = false;
+
+            var rules = _ruleFactory.CreateAnonymizationDicomRule(_configurationManager.Configuration.RuleContent);
+            foreach (var rule in rules)
+            {
+                rule.Handle(dataset, context);
+                _logger.LogDebug($"Successfully handled rule {rule.Description}.");
+            }
+
+            ValidateOutput(dataset);
+        }
+
+        private void ValidateInput(DicomDataset dataset)
+        {
+            if (_anonymizerSettings.ValidateInput)
+            {
+                dataset.Validate();
+            }
+        }
+
+        private void ValidateOutput(DicomDataset dataset)
+        {
+            if (_anonymizerSettings.ValidateOutput)
+            {
+                dataset.Validate();
+            }
+        }
+
+        private ProcessContext InitContext(DicomDataset dataset)
```

Comment:
>InitContext [](http://example.com/codeflow?start=31&length=11)

should you call EnsureArg.IsNotNull to check the parameter first?

Reply:
Since it is a private function, and we have checked dataset in not null outside.